fontbe/src/kern.rs

```diff
@@ -0,0 +1,138 @@
+//! Generates a Kerning datastructures to be fed to fea-rs
+
+use std::collections::{BTreeMap, HashSet};
+
+use fea_rs::{compile::PairPosBuilder, GlyphSet};
+use font_types::GlyphId;
+use fontdrasil::orchestration::{Access, Work};
+use write_fonts::tables::gpos::ValueRecord;
+
+use crate::{
+    error::Error,
+    features::{create_glyphmap, resolve_variable_metric},
+    orchestration::{AnyWorkId, BeWork, Context, Kerning, WorkId},
+};
+use fontir::{ir::KernParticipant, orchestration::WorkId as FeWorkId};
+
+#[derive(Debug)]
+struct KerningWork {}
+
+pub fn create_kerning_work() -> Box<BeWork> {
+    Box::new(KerningWork {})
+}
+
+impl Work<Context, AnyWorkId, Error> for KerningWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Kerning.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        Access::Set(HashSet::from([
+            FeWorkId::StaticMetadata.into(),
+            FeWorkId::Kerning.into(),
+            FeWorkId::GlyphOrder.into(),
+        ]))
+    }
+
+    /// Generate kerning data structures.
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let static_metadata = context.ir.static_metadata.get();
+        let glyph_order = context.ir.glyph_order.get();
+        let kerning = context.ir.kerning.get();
+        let gid = |name| {
+            glyph_order
+                .glyph_id(name)
+                .map(|gid| GlyphId::new(gid as u16))
+                .ok_or_else(|| Error::MissingGlyphId(name.clone()))
+        };
+
+        let fears_glyph_map = create_glyphmap(&glyph_order);
+
+        // convert the groups stored in the Kerning object into the glyph classes
+        // expected by fea-rs:
+        let glyph_classes = kerning
+            .groups
+            .iter()
+            .map(|(class_name, glyph_set)| {
+                let glyph_class: GlyphSet = glyph_set
+                    .iter()
+                    .map(|name| GlyphId::new(glyph_order.glyph_id(name).unwrap_or(0) as u16))
+                    .collect();
+                (class_name, glyph_class)
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        let mut ppos_subtables = PairPosBuilder::default();
+
+        // now for each kerning entry, directly add a rule to the builder:
+        for ((left, right), values) in &kerning.kerns {
+            let (default_value, deltas) = resolve_variable_metric(&static_metadata, values)?;
+
+            let mut x_adv_record = ValueRecord::new().with_x_advance(default_value);
+            let empty_record = ValueRecord::default();
+
+            if !deltas.is_empty() {
+                // COLIN HOW DO I DO THIS WITHOUT A BUILDER?
+                let var_idx = builder.add_deltas(deltas);
```

Comment:
@cmyr I'm stuck here. IIUC it's all based on temporary ids until the actual compile so maybe I can fabricate one?

Reply:
hmmm you do need the real builder to get an ID, currently, but I think I could make it so that you can have two separate builders and merge them? So maybe for now you should just create your own builder and then I'll figure out the merging after.

Reply:
Per IM & VC, might need to produce non-fea-rs structures until such time as we have the ability to accumulate deltas without var indices. That's https://github.com/googlefonts/fontc/issues/571.